Issue #17 : Enable deleting settings in snapshot wp-admin/post page.

kienstra opened the pull request:
**Request For Code Review**

Hi @westonruter,
Could you please review this pull request for Issue #17?
- [73c5b28](https://github.com/xwp/wp-customize-snapshots/commit/73c5b28201f8f951f8c3e938405486cfb6f9a40e) : [Create link](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-36a80da8dbba0bdbe840617270bb1182R365) `'Remove setting'`, which toggles to `'Restore setting'` [on click](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-19c3ccdc306cf3ee7efd933155ccae35R18).
  It also [sets a hidden input field](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-19c3ccdc306cf3ee7efd933155ccae35R40) for each setting to be removed.
  On saving, these appear in `$_REQUEST[ 'customize_snapshot_remove_settings' ]`.
  Filter post content  [on content_save_pre](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-36a80da8dbba0bdbe840617270bb1182R117) when this is saved.
  If `$_REQUEST` has settings to be removed, [filter them out](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-36a80da8dbba0bdbe840617270bb1182R722).
  Also, add [simple styling](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-96bf05ff71d50822b5e71fea81a765e1R1) of this 'Remove setting' link.
  It [appears red](https://github.com/xwp/wp-customize-snapshots/compare/feature/allow-deleting-settings?expand=1#diff-96bf05ff71d50822b5e71fea81a765e1R4) when it has the text "Remove setting, and blue otherwise.

Commit:
Issue #17 : Enable deleting settings in snapshot wp-admin/post page.

Create link 'Remove setting,' which toggles to 'Restore setting' on click.
It also sets a hidden text field for each setting to be removed.
On saving, these appear in [ 'customize_snapshot_remove_settings' ].
Filter post content  on 'content_save_pre,' when this is saved.
If  has settings to be removed, filter them out.
Also, add simple styling of this 'Remove setting' link.
It appears red when setto 'remove,' and blue when set to 'restore.'

Commit:
Issue #17 - Fix ESLint errors, including removing extra parentheses.

Commit:
Issue #17 : Correct ESLint error by removing parameter .

Commit:
Issue #17 : Fix failing unit test: property of non-object error.

Before evaluating ->post_type, pass it to isset().
Short-circuit the conditional if it's not set.

westonruter commented on php/class-post-type.php:
```diff
@@ -675,4 +678,53 @@ public function hide_disabled_publishing_actions( $post ) {
+			&&
+			wp_verify_nonce( $_REQUEST[ static::SLUG ], static::SLUG . '_settings' )
+			&&
+			! ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE )
```
We should be using the new `wp_doing_ajax` function if it is available, so this can be replaced with:

``` php
! ( function_exists( 'wp_doing_ajax' ) ? wp_doing_ajax() : defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE )
```

This will likely yield benefits for doing unit testing without having to be in a separate Ajax unit test, since the return value of `wp_doing_ajax()` can be filtered as opposed to being a fixed constant.

westonruter replied:
Oops. 😊 I misread the constant here. Nevermind 😄

westonruter committed:
Merge branch 'develop' of https://github.com/xwp/wp-customize-snapshots into feature/allow-deleting-settings

westonruter committed:
Chose line-through to mark removed settings; use unabbreviated class names

westonruter commented on php/class-post-type.php:
```diff
@@ -675,4 +678,53 @@ public function hide_disabled_publishing_actions( $post ) {
+		}
+
+		$setting_ids_to_unset = $_REQUEST[ $key_for_settings ];
+		$data = json_decode( $post->post_content );
```
Instead of getting `$post->post_content` should it not be using the `$content` that is already provided?

kienstra replied:
That's a good point. [This commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/f76b4a6dba05988f250582dcd87db61902ac3ec0) corrects that.

westonruter commented on php/class-post-type.php:
```diff
@@ -675,4 +678,53 @@ public function hide_disabled_publishing_actions( $post ) {
+			&&
+			( static::SLUG === $post->post_type )
+			&&
+			! empty( $_REQUEST[ $key_for_settings ] )
```
Should this be checked to see if it `is_array`?

kienstra replied:
That's a good idea. This [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/d1487c8c233a45e2ac3a7db32780dd6a45f17d1f) adds an `is_array` call to `$should_fliter_content`.

westonruter commented on php/class-customize-snapshot-manager.php:
```diff
@@ -745,6 +746,21 @@ public function enqueue_frontend_scripts() {
+	public function enqueue_admin_scripts() {
+		global $post;
+		$handle = 'customize-snapshots-admin';
+		if ( isset( $post->post_type ) && ( Post_Type::SLUG === $post->post_type ) && ( 'publish' !== $post->post_status ) ) {
```
Should this be looking at the `$hook` arg passed in? Or should it be checking what is returned by `get_current_screen()`?

kienstra replied:
Hi @weston, it looks like checking the `$hook` argument in this [admin_enqueue_scripts](https://codex.wordpress.org/Plugin_API/Action_Reference/admin_enqueue_scripts) callback should be enough.
This [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/68fa5ae929700ab44b31f171d69d09f97074610b) adds an expression to the conditional:
`(  'post.php' === $hook )`
The [WP_Screen object](https://codex.wordpress.org/Class_Reference/WP_Screen) that `get_current_screen()` returns has some useful properties, including:

```
id = "custom_snapshot",
base = "post",
post_type = "customize_snapshot"
```

But the conditional already checks for most of these:
`if ( ( 'post.php' === $hook ) && isset( $post->post_type ) && ( Post_Type::SLUG === $post->post_type ) && ( 'publish' !== $post->post_status ) ) {`

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+		var $link, linkText, linkActions, dataSlug, initializeLink;
+
+		$link = $( '.cs-toggle-action' );
+		linkText = [ 'Remove setting', 'Restore setting' ];
```
These strings need to be translated, and exported from PHP into JS.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+		var $link, linkText, linkActions, dataSlug, initializeLink;
+
+		$link = $( '.snapshot-toggle-setting-removal' );
+		linkText = [ 'Remove setting', 'Restore setting' ];
```
These strings need to be translated, and exported from PHP into JS.

kienstra replied:
That's a good point. [This commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/35d5bca1f03a9fb5cd1aef66a355b91dd30eba33#diff-36a80da8dbba0bdbe840617270bb1182R365) outputs the translated strings via PHP.

westonruter commented on php/class-post-type.php:
```diff
@@ -360,6 +362,7 @@ public function render_data_metabox( $post ) {
 			echo '<li>';
 			echo '<details open>';
 			echo '<summary><code>' . esc_html( $setting_id ) . '</code> ';
+			echo '<a href="#" id="' . esc_attr( $setting_id ) . '" class="snapshot-toggle-setting-removal remove"></a>';
```
I suggest you output the actual strings here. You can do like:

``` php
echo '<a href="#" id="' . esc_attr( $setting_id ) . '" data-text-restore="<?php esc_attr_e( 'Restore setting', 'customize-snapshots' ) ?>" class="snapshot-toggle-setting-removal remove"><?php esc_html_e( 'Remove setting', 'customize-snapshots' ) ?></a>';
```

kienstra replied:
Thanks, @westonruter. [This line](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/35d5bca1f03a9fb5cd1aef66a355b91dd30eba33#diff-36a80da8dbba0bdbe840617270bb1182R365) in a new commit outputs the strings, and [this function](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/35d5bca1f03a9fb5cd1aef66a355b91dd30eba33#diff-19c3ccdc306cf3ee7efd933155ccae35R59) toggles between the strings in JavaScript.

westonruter commented on php/class-post-type.php:
```diff
@@ -360,6 +362,7 @@ public function render_data_metabox( $post ) {
 			echo '<li>';
 			echo '<details open>';
 			echo '<summary><code>' . esc_html( $setting_id ) . '</code> ';
+			echo '<a href="#" id="' . esc_attr( $setting_id ) . '" class="snapshot-toggle-setting-removal remove"></a>';
```
Then you can toggle between these strings in the JS.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+			clickedLinkAction = $( this ).data( dataSlug );
+			settingId = $( this ).attr( 'id' );
+
+			this.isLinkSetToRemoveSetting = function() {
```
Humm. You're adding these functions to the jQuery object itself. Maybe better to create a `component` object that you attach the functions to, and then call its methods in a click handler instead? You can see a pattern for how to create a `component` in https://github.com/xwp/wp-customize-snapshots/blob/develop/js/customize-snapshots.js and see also how it gets exported to global scope so that the component can be interacted with by other scripts, or even by you on the console.

kienstra replied:
Thanks, @westonruter. This [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/83ad0e9ee2983c6fc7e35c6c55836e9c4242f3f0) attaches the functions to `component`. But I still need to export `component` object.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+					.addClass( 'snapshot-setting-removed' );
+			};
+
+			this.constructHiddenInputWithValue = function() {
```
Better for this function to be more of a pure function that doesn't rely on its closure scope. So it can be at the top level as `component.constructHiddenInputWithValue()` and it can take a `settingId` as its argument. It is much easier to test the function like this, and there is only ever one copy, whereas here there is a copy for each time the user clicks.

kienstra replied:
Thanks, this [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/83ad0e9ee2983c6fc7e35c6c55836e9c4242f3f0) moves `constructHiddenInputWithValue` into the `component` object.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+
+		initializeLink();
+
+		$link.on( 'click', function( event ) {
```
The function body should be refactored, to make the logic for the functions to not rely on `this` but be passed the context as an explicit argument. So `hideSettingAndChangeLinkText` can take an argument for the setting `details` jQuery object, for example, instead of relying on the inherited `this` context.

kienstra replied:
Thanks, @westonruter. That's a good idea to refactor the functions. This [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/83ad0e9ee2983c6fc7e35c6c55836e9c4242f3f0) does that. Sorry for the delay.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+		initializeLink = function() {
+			$link.text( linkText[ 0 ] )
+				.data( dataSlug, linkActions[ 0 ] );
+		};
```
This can be eliminated in favor of putting the translated text strings inline in the rendered PHP.

kienstra replied:
Thanks, [this commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/35d5bca1f03a9fb5cd1aef66a355b91dd30eba33#diff-36a80da8dbba0bdbe840617270bb1182R363) outputs the translated strings with PHP.

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,73 @@
+			};
+
+			this.removeHiddenInputWithValue = function() {
+				$( 'input[value="' + settingId + '"]' ).remove();
```
Also make sure that the `name` is `customize_snapshot_remove_settings[]` right?

kienstra replied:
That's a good point, @westonruter. This [commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/83ad0e9ee2983c6fc7e35c6c55836e9c4242f3f0#diff-19c3ccdc306cf3ee7efd933155ccae35R73) ensures that the `name` is `customize_snapshot_remove_settings[]`.

westonruter commented on php/class-post-type.php:
```diff
@@ -114,6 +114,7 @@ public function register() {
 		add_filter( 'display_post_states', array( $this, 'display_post_states' ), 10, 2 );
 		add_action( 'admin_notices', array( $this, 'show_publish_error_admin_notice' ) );
 		add_action( 'post_submitbox_minor_actions', array( $this, 'hide_disabled_publishing_actions' ) );
+		add_filter( 'content_save_pre', array( $this, 'filter_out_settings_if_removed_in_metabox' ), 10 );
```
Good idea to use `content_save_pre`.

westonruter commented on php/class-post-type.php:
```diff
@@ -675,4 +678,53 @@ public function hide_disabled_publishing_actions( $post ) {
+		$setting_ids_to_unset = $_REQUEST[ $key_for_settings ];
+		$data = json_decode( $post->post_content );
+		foreach ( $setting_ids_to_unset as $setting_id ) {
+			unset( $data->{ $setting_id } );
```
The second `$associative` arg should be passed to `json_decode()` above so that it will be an array so that the familiar bracket notation can be used as opposed to this somewhat-unusual object brace notation.

kienstra replied:
Thanks, [this commit](https://github.com/xwp/wp-customize-snapshots/pull/84/commits/92f14c04d726faa219e31350ad2ad15b5730566e) uses the second argument of `true` to return an array.

westonruter commented:
@kienstra the functionality works well! Beyond my inline feedback, something else that should be done is a unit test on `filter_out_settings_if_removed_in_metabox`.

Commit:
Issue #17 : Refactor customize-snapshots-admin.js methods.

Instead of attaching methods to 'on' event handler,
Attach them to module-scope components object.
Haven't yet implemented a way to export them, as Weston suggested.

Commit:
Issue #17 : Change logic for toggling link text.

Using Weston's snippet, output link title in PHP, with localization.
And attach localized text 'Restore settting' to the link's data-text-restore attribute.
On clicking the link, call method changeLinkText.
This toggles the title to the other value stored in the data attribute.
And stores the old title in the data attribute, for the next time it's clicked.

Commit:
Issue #17 : Remove need for object brace notation.

At Weston's suggestion, pass second argument to json_decode.
This will then return an array, instead of an object.
So access the array value with $data[ $setting_id ].
Instead of previous object access with $data->{ $setting_id }.

Commit:
Issue #17 : Use $content passed to filter, instead of $post->post_content.

Also, wp_unslash() the $content.
Otherwise, the json_decode function will return null.

Commit:
Issue #17 : Add is_array() to $should_filter_content conditional.

Before iterating through the settings to unset,
Ensure that the settings are in an array, as expected.

Commit:
Issue #17 : In enqueuing, verify that $hook argument is 'post.php'.

kienstra commented:
**Thanks For Waiting**
_I Still Need To Write PHPUnit Tests_

Hi @westonruter,
Sorry for the delay in applying your feedback here. I addressed all of your comments above, but I still need to write PHPUnit tests.

westonruter commented:
@kienstra no apology will be accepted! It is I who probably should apologize for being slow to review PRs 😞

PatelUtkarsh commented on php/class-post-type.php:
```diff
@@ -360,6 +362,7 @@ public function render_data_metabox( $post ) {
 			echo '<li>';
 			echo '<details open>';
 			echo '<summary><code>' . esc_html( $setting_id ) . '</code> ';
+			echo '<a href="#" id="' . esc_attr( $setting_id ) . '" data-text-restore="' . esc_attr( 'Restore setting', 'customize-snapshots' ) . '" class="snapshot-toggle-setting-removal remove">' . esc_html( 'Remove setting', 'customize-snapshots' ) . '</a>';
```
Use translation function `esc_attr__` and `esc_html__`  instade of `esc_attr( 'Restore setting', 'customize-snapshots' )` and `esc_html( 'Remove setting', 'customize-snapshots' )`

PatelUtkarsh committed:
Merge 'develop' in 'feature/allow-deleting-settings'

PatelUtkarsh commented:
In e5accee I've resolved conflict with develop.

PatelUtkarsh commented:
I am picking this up as discussed with @jeffpaul and @kienstra.

PatelUtkarsh committed:
Use translation function for escaping

PatelUtkarsh committed:
Add unit-test for delete setting

westonruter committed:
Replace testing trunk with testing latest due to lack of changesets compat
